core/src/main/java/ai/timefold/solver/core/impl/phase/custom/DefaultCustomPhase.java

```diff
@@ -1,27 +1,48 @@
 package ai.timefold.solver.core.impl.phase.custom;
 
+import java.util.Iterator;
 import java.util.List;
 
 import ai.timefold.solver.core.api.domain.solution.PlanningSolution;
 import ai.timefold.solver.core.impl.phase.AbstractPhase;
+import ai.timefold.solver.core.impl.phase.PossiblyInitializingPhase;
 import ai.timefold.solver.core.impl.phase.custom.scope.CustomPhaseScope;
 import ai.timefold.solver.core.impl.phase.custom.scope.CustomStepScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
 import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 import ai.timefold.solver.core.impl.solver.scope.SolverScope;
 import ai.timefold.solver.core.impl.solver.termination.Termination;
 
+import org.jspecify.annotations.NullMarked;
+
 /**
  * Default implementation of {@link CustomPhase}.
  *
  * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
  */
-final class DefaultCustomPhase<Solution_> extends AbstractPhase<Solution_> implements CustomPhase<Solution_> {
+@NullMarked
+public final class DefaultCustomPhase<Solution_>
```

Comment:
This class shares a logic similar to that of the construction phase. We can consider adding a new abstract phase, `AbstractPossiblyInitializingPhase`, that extends `AbstractPhase` and implements `PossiblyInitializingPhase`. The new class can specify the relevant attributes and implement the related phase events.